pkg/k8s/cnpstatus.go

```diff
@@ -0,0 +1,345 @@
+func (c *CNPStatusEventHandler) StopStatusHandler(cnp *types.SlimCNP) {
+	cnpKey := getKeyFromObjectMeta(cnp.ObjectMeta)
+	prefix := path.Join(CNPStatusesPath, cnpKey)
+	err := kvstore.DeletePrefix(prefix)
+	if err != nil {
+		log.WithError(err).WithField("prefix", prefix).Warning("error deleting prefix from kvstore")
+	}
+	c.eventMap.delete(cnpKey)
+
+}
+
+func (c *CNPStatusEventHandler) runStatusHandler(cnpKey string, cnp *types.SlimCNP, nodeStatusUpdater *NodeStatusUpdater) {
+	namespace := cnp.Namespace
+	name := cnp.Name
+	nodeStatusMap := make(map[string]cilium_v2.CiliumNetworkPolicyNodeStatus)
+
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.K8sNamespace:            namespace,
+		logfields.CiliumNetworkPolicyName: name,
+	})
+
+	scopedLog.Debug("started status handler")
+
+	// Iterate over the shared-store first. We may have received events for this
+	// CNP in the key-value store from nodes which received and processed this
+	// CNP and sent status updates for it before the watcher which updates this
+	// `CNPStatusEventHandler` did. Given that we have the shared store which
+	// caches all keys / values from the kvstore, we iterate and collect said
+	// events. Given that this function is called after we have updated the
+	// `eventMap` for this `CNPStatusEventHandler`, subsequent key updates from
+	// the kvstore are guaranteed to be sent on the channel in the
+	// `nodeStatusUpdater`, which we will receive in the for-loop below.
+	sharedKeys := c.cnpStore.SharedKeysMap()
+	for keyName, storeKey := range sharedKeys {
+		// Look for any key which matches this CNP.
+		if strings.HasPrefix(keyName, cnpKey) {
+			cnpns, ok := storeKey.(*CNPNSWithMeta)
+			if !ok {
+				scopedLog.Errorf("received unexpected type mapping to key %s in cnp shared store: %T", keyName, storeKey)
+				continue
+			}
+			// extract nodeName from keyName
+			nodeStatusMap[cnpns.Node] = cnpns.CiliumNetworkPolicyNodeStatus
+		}
+	}
+	for {
+		// If nothing to update, wait until we have something to update.
+		if len(nodeStatusMap) == 0 {
+
+			// Wait for one status update to occur.
+			select {
+			case <-nodeStatusUpdater.stopChan:
+				return
+			case ev, ok := <-nodeStatusUpdater.updateChan:
+				if !ok {
+					return
+				}
+				nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
+			}
+		}
+
+		// Allow for a bunch of different node status updates to come before
+		// we break out to avoid jitter in updates across the cluster
+		// to affect batching on our end.
+		limit := time.After(c.updateInterval)
+
+		// Collect any other events that have come in, but bail out after the
+		// above limit is hit so that we can send the updates we have received.
+	Loop:
+		for {
+			select {
+			case <-nodeStatusUpdater.stopChan:
+				return
+			case <-limit:
+				break Loop
+			case ev, ok := <-nodeStatusUpdater.updateChan:
+				if !ok {
+					return
+				}
+				nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
+			}
+		}
```

Comment:
```golang
	for {
		limit := time.After(c.updateInterval)

		// Collect any other events that have come in, but bail out after the
		// above limit is hit so that we can send the updates we have received.
	Loop:
		for {
			select {
			case <-nodeStatusUpdater.stopChan:
				return
			case <-limit:
				if len(nodeStatusMap) == 0 {
					// If nothing to update, wait until we have something to update.
					limit = nil
					continue
				}
				break Loop
			case ev, ok := <-nodeStatusUpdater.updateChan:
				if !ok {
					return
				}
				nodeStatusMap[ev.node] = *ev.CiliumNetworkPolicyNodeStatus
			}
		}
```